### PR TITLE
Determinism fixes and full output precision

### PR DESCRIPTION
**Pull request by nh2**
Hey,

here's a set of fixes we carry locally that we'd like to upstream.

The key points are

* determinism (following up on #36, #37, #38), in this case by using a fixed seed for shuffling
* outputting accuracy as a field in the output `.ply` files
* fixing a division by zero
* enabling full output precision of floating point values being emitted (see commit message for an example)

For the last point, it fixes expected output like this:

image

degenerating into output like this when the object is not close to the origin:

image

Close-up:

image

### CODE REVIEW

**Comment by nh2** [comment]
I can split up the commits into separate PRs if you don't want all of them batched, but I thought I'd first show them all in case they are desired.

File: program/base/cmvs/bundle.cc [hunk 1]
```diff
@@ -662,7 +663,7 @@ void Cbundle::findPNeighbors(sfcnn<const float*, 3, float>& tree,
 }
 
 void Cbundle::mergeSfMPThread(void) {
-  const int tenth = (int)m_coords.size() / 10;
```
Comment (pmoulon) [inline program/base/cmvs/bundle.cc]: could we use std::max(), I prefer the addition of the `std::` to ensure we are using the standard one

File: program/base/cmvs/bundle.cc [hunk 2]
```diff
@@ -758,7 +759,8 @@ void Cbundle::mergeSfMP(void) {
     order.resize(cpnum);
     for (int p = 0; p < cpnum; ++p)
       order[p] = p;
-    random_shuffle(order.begin(), order.end());
+    std::mt19937 gen(42);
```
Comment (pmoulon) [inline program/base/cmvs/bundle.cc]: Why 42? Can we made this a static const variable in the file?

File: program/base/pmvs/patchOrganizerS.cc [hunk 3]
```diff
@@ -91,6 +93,11 @@ void CpatchOrganizerS::writePatches2(const std::string prefix, bool bExportPLY,
     sprintf(buffer, "%s.patch", prefix.c_str());
     ofstream ofstr;
     ofstr.open(buffer);
+
+    // Set output to full precision to be able to fully round-trip
+    // the text representation with what we have in memory.
+    ofstr << std::setprecision(std::numeric_limits<double>::max_digits10);
```
Comment (pmoulon) [inline program/base/pmvs/patchOrganizerS.cc]: Like this, I'm doing it in OpenMVG PLY export, but did not think about adding it here ;-)

File: program/base/pmvs/patchOrganizerS.cc [hunk 4]
```diff
@@ -708,6 +720,7 @@ void CpatchOrganizerS::writePLY(const std::vector<Ppatch>& patches,
        << "property uchar diffuse_red" << '\n'
        << "property uchar diffuse_green" << '\n'
        << "property uchar diffuse_blue" << '\n'
+       << "property float quality" << '\n'
```
Comment (pmoulon) [inline program/base/pmvs/patchOrganizerS.cc]: Do we need to export this new field?

File: program/base/pmvs/seed.cc [hunk 5]
```diff
@@ -55,7 +56,8 @@ void Cseed::run(void) {
   for (int i = 0; i < m_fm.m_tnum; ++i)
     vitmp.push_back(i);
```
Comment (pmoulon) [inline program/base/pmvs/seed.cc]: Nit: We could replace this with https://en.cppreference.com/w/cpp/algorithm/iota

File: program/base/pmvs/seed.cc [hunk 6]
```diff
@@ -55,7 +56,8 @@ void Cseed::run(void) {
...
 
-  random_shuffle(vitmp.begin(), vitmp.end());
+  std::mt19937 gen(42);
```
Comment (pmoulon) [inline program/base/pmvs/seed.cc]: Would be nice to have the seed number define once and customizable

File: program/thirdParty/graclus1.2/metisLib/util.c [hunk 7]
```diff
@@ -106,7 +106,7 @@ void errexit(char *f_str,...)
 
   sprintf(out2, "Error! %s", out1);
 
-  fprintf(stdout, out2);
+  fputs(stdout, out2);
```
Comment (pmoulon) [inline program/thirdParty/graclus1.2/metisLib/util.c]: What is the motivation behind this change?

Comment (nh2) [reply program/thirdParty/graclus1.2/metisLib/util.c]: Please see the commit message.

Copying the explanation here for convenience: It fixes a warning:

```
program/thirdParty/graclus1.2/metisLib/util.c:109:19: warning: format not a string literal and no format arguments [-Wformat-security]
   fprintf(stdout, out2);
                   ^
```

Comment (nh2) [reply program/thirdParty/graclus1.2/metisLib/util.c]: Fixing a warning, see commit messag.

File: program/base/cmvs/bundle.cc [hunk 8]
```diff
@@ -662,7 +663,7 @@ void Cbundle::findPNeighbors(sfcnn<const float*, 3, float>& tree,
 }
 
 void Cbundle::mergeSfMPThread(void) {
-  const int tenth = (int)m_coords.size() / 10;
+  const int tenth = max(1, (int)m_coords.size() / 10);
```
Comment (pmoulon) [inline program/base/cmvs/bundle.cc]: Can we add the std:: here too?

Comment (nh2) [reply program/base/cmvs/bundle.cc]: Ah, my bad. Fixed.

Comment (nh2) [reply program/base/cmvs/bundle.cc]: I've also updated the last remaining plain `max` in this file that was there before this PR.

**Comment by pmoulon** [comment]
Thx @nh2
